src/tournaments_watcher/ema.rs

```diff
@@ -0,0 +1,277 @@
+    check_rcr_ruleset(entries_new);
+
+    let mut old_entries_map: HashMap<String, &TournamentEntry> = HashMap::new();
+    for entry in unique_entries(entries_old.get()) {
+        old_entries_map.insert(entry.name.clone(), entry);
+    }
+
+    let mut statuses = Vec::new();
+    for new_entry in unique_entries(entries_new.get()) {
+        let old_entry = old_entries_map.get(&new_entry.name).copied();
+        let status = TournamentStatus::for_entries(old_entry, new_entry);
+        if let Some(status) = status {
+            statuses.push(status);
+        }
+    }
+
+    TournamentStatuses(statuses)
+}
+
+fn check_rcr_ruleset(entries_old: &Tournaments) {
+    assert!(
+        entries_old
+            .get()
+            .iter()
+            .all(|entry| entry.rules == RCR_RULES_NAME),
+        "Only a Riichi ruleset is supported, please filter your data first"
+    );
+}
+
+fn unique_entries<'a, T>(entries: T) -> impl Iterator<Item = &'a TournamentEntry>
+where
+    T: IntoIterator<Item = &'a TournamentEntry>,
+{
+    entries.into_iter().unique_by(|entry| &entry.name)
+}
+
+#[must_use]
+fn get_diff_option(value_old: &str, value_new: &str) -> Option<String> {
+    if value_old == value_new {
+        None
+    } else {
+        Some(value_new.to_owned())
+    }
+}
+
+pub fn parse_tournaments(body: &str) -> anyhow::Result<Tournaments> {
+    let html = Html::parse_document(body);
+    let table = select_one!(".Tableau_CertifiedTournament .TCTT_lignes", html)?;
+    let rows: Vec<_> = select_all!("div", table).collect();
+
+    let mut last_header = String::new();
+    let mut entries = Vec::new();
+    for row in rows {
+        let cells: Vec<_> = select_all!("p", row).collect();
+        if is_header(&cells) {
+            last_header = first_nonempty_text(&cells[0])?.to_owned();
+        } else {
+            let entry = make_entry(&mut last_header, cells)?;
+            entries.push(entry);
+        }
+    }
+
+    Ok(Tournaments(entries))
+}
+
+#[must_use]
+fn is_header(cells: &[ElementRef]) -> bool {
+    let first_cell = cells[0];
+    let mut cell_classes = first_cell.value().classes();
+    cell_classes.any(|class_name| class_name.starts_with(HEADER_CLASS_PREFIX))
+}
+
+fn make_entry(last_header: &mut String, cells: Vec<ElementRef>) -> anyhow::Result<TournamentEntry> {
+    let url = if let Some(element) = select_all!("a", cells[0]).next() {
+        element
+            .value()
+            .attr("href")
+            .ok_or(anyhow!("<a> element does not contain a link"))?
+            .to_owned()
+    } else {
+        "".to_owned()
+    };
+    let texts = cells.iter().map(cell_text).collect::<Vec<_>>();
+
+    let entry = TournamentEntry {
+        name: texts[0].clone(),
+        url,
+        rules: texts[1].clone(),
+        date: format!("{} {}", texts[2], last_header),
+        place: texts[3].clone(),
+        approval_status: texts[4].clone(),
+        results_status: texts[5].clone(),
```

Comment:
do we want to do some basic bound check, e.g. that `texts`' length is >= 6, to avoid panics here?

Reply:
Let's.